# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Comprehensive Bazel BUILD files patch for WasmEdge 0.16.1
# This patch adds BUILD.bazel files throughout the WasmEdge directory structure
# for a native Bazel build using rules_cc instead of rules_foreign_cc.
#
# Based on https://github.com/mmorel-35/WasmEdge/pull/1 but adapted for
# official WasmEdge 0.16.1 sources with C++20 compatibility fixes.

--- a/include/experimental/span.hpp
+++ b/include/experimental/span.hpp
@@ -37,6 +37,8 @@ template <class T, class = void> struct defined_to_address : false_type {};
 template <class T>
 struct defined_to_address<T, void_t<decltype(pointer_traits<T>::to_address)>>
     : true_type {};
+
+#if !defined(__cpp_lib_to_address) || __cpp_lib_to_address < 201711L
 } // namespace detail
 
 template <class T> constexpr T *to_address(T *p) noexcept {
@@ -51,6 +53,13 @@ template <class T> constexpr auto to_address(const T &p) noexcept {
     return to_address(p.operator->());
   }
 }
+#else
+} // namespace detail
+
+// C++20 or later with std::to_address available - use it from std
+using std::to_address;
+
+#endif // !defined(__cpp_lib_to_address) || __cpp_lib_to_address < 201711L
 
 namespace detail {
 
--- /dev/null
+++ b/BUILD.bazel
@@ -0,0 +1,37 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+#
+# Root BUILD file for WasmEdge
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+# Main WasmEdge library - aggregates all components
+cc_library(
+    name = "wasmedge_lib",
+    deps = [
+        "//lib/api:wasmedge",
+    ],
+)
+
+# Interpreter-only build (no AOT/LLVM)
+cc_library(
+    name = "wasmedge_interpreter",
+    deps = [
+        "//lib/common",
+        "//lib/loader",
+        "//lib/validator",
+        "//lib/executor",
+        "//lib/vm",
+    ],
+)
+
+# Exports for external use
+exports_files([
+    "include/api/wasmedge/wasmedge.h",
+    "include/api/wasmedge/version.h.in",
+])
+
+licenses(["notice"])  # Apache 2.0
+
--- /dev/null
+++ b/include/BUILD.bazel
@@ -0,0 +1,102 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+#
+# Generated headers for WasmEdge
+
+package(default_visibility = ["//visibility:public"])
+
+# Generate api/wasmedge/version.h from template
+genrule(
+    name = "generate_api_version_header",
+    srcs = ["api/wasmedge/version.h.in"],
+    outs = ["api/wasmedge/version.h"],
+    cmd = """
+        sed -e 's/$${CPACK_PACKAGE_VERSION}/0.16.1/g' \\
+            -e 's/$${WASMEDGE_VERSION_MAJOR}/0/g' \\
+            -e 's/$${WASMEDGE_VERSION_MINOR}/16/g' \\
+            -e 's/$${WASMEDGE_VERSION_PATCH}/1/g' \\
+            -e 's/$${WASMEDGE_API_VERSION}/4/g' \\
+            $(location api/wasmedge/version.h.in) > $@
+    """,
+)
+
+# Generate common/config.h from template
+genrule(
+    name = "generate_config_header",
+    srcs = ["common/config.h.in"],
+    outs = ["common/config.h"],
+    cmd = """
+        cat > $@ << 'EOF'
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2019-2024 Second State INC
+// Generated by Bazel build
+
+#pragma once
+
+#include <string_view>
+
+namespace WasmEdge {
+
+using namespace std::literals::string_view_literals;
+
+static inline constexpr std::string_view kCacheRoot [[maybe_unused]] = "/var/cache/wasmedge"sv;
+
+#define HAVE_MMAP 1
+#define HAVE_PWD_H 1
+
+} // namespace WasmEdge
+EOF
+    """,
+)
+
+# Generate common/version.h from template
+genrule(
+    name = "generate_common_version_header",
+    srcs = ["common/version.h.in"],
+    outs = ["common/version.h"],
+    cmd = """
+        cat > $@ << 'EOF'
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2019-2024 Second State INC
+// Generated by Bazel build
+
+#pragma once
+
+#include <string_view>
+
+namespace WasmEdge {
+
+using namespace std::literals::string_view_literals;
+
+static inline std::string_view kVersionString [[maybe_unused]] = "0.16.1"sv;
+static inline std::string_view kGlobalPluginDir [[maybe_unused]] = "/usr/local/lib/wasmedge"sv;
+static inline constexpr const uint32_t kPluginCurrentAPIVersion [[maybe_unused]] = 4;
+
+} // namespace WasmEdge
+EOF
+    """,
+)
+
+# Copy enum headers from common/ to api/wasmedge/ as CMake does
+genrule(
+    name = "copy_enum_headers",
+    srcs = [
+        "common/enum.inc",
+        "common/enum_configure.h",
+        "common/enum_errcode.h",
+        "common/enum_types.h",
+    ],
+    outs = [
+        "api/wasmedge/enum.inc",
+        "api/wasmedge/enum_configure.h",
+        "api/wasmedge/enum_errcode.h",
+        "api/wasmedge/enum_types.h",
+    ],
+    cmd = """
+        cp $(location common/enum.inc) $(location api/wasmedge/enum.inc)
+        cp $(location common/enum_configure.h) $(location api/wasmedge/enum_configure.h)
+        cp $(location common/enum_errcode.h) $(location api/wasmedge/enum_errcode.h)
+        cp $(location common/enum_types.h) $(location api/wasmedge/enum_types.h)
+    """,
+)
+
--- /dev/null
+++ b/lib/BUILD.bazel
@@ -0,0 +1,10 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+#
+# Root lib BUILD file
+
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])  # Apache 2.0
+
--- /dev/null
+++ b/lib/common/BUILD.bazel
@@ -0,0 +1,32 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "common",
+    srcs = [
+        "errinfo.cpp",
+        "hash.cpp",
+        "hexstr.cpp",
+        "spdlog.cpp",
+    ],
+    hdrs = glob([
+        "../../include/common/*.h",
+        "../../include/common/*.hpp",
+        "../../include/common/*.inc",
+        "../../include/experimental/*.hpp",
+    ]) + [
+        "//include:generate_config_header",
+        "//include:generate_common_version_header",
+    ],
+    includes = [
+        "../../include",
+    ],
+    deps = [
+        "@spdlog",
+    ],
+)
+
--- /dev/null
+++ b/lib/loader/BUILD.bazel
@@ -0,0 +1,33 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "loader",
+    srcs = glob([
+        "*.cpp",
+        "filemgr/*.cpp",
+        "ast/*.cpp",
+        "serialize/*.cpp",
+    ]),
+    hdrs = glob([
+        "../../include/loader/*.h",
+        "../../include/loader/filemgr/*.h",
+        "../../include/aot/*.h",
+        "../../include/ast/*.h",
+        "../../include/ast/component/*.h",
+    ]),
+    includes = [
+        "../../include",
+    ],
+    deps = [
+        "//lib/common",
+        "//lib/system",
+        "@simdjson",
+    ],
+)
+
--- /dev/null
+++ b/lib/system/BUILD.bazel
@@ -0,0 +1,37 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "system",
+    srcs = [
+        "allocator.cpp",
+        "fault.cpp",
+        "mmap.cpp",
+        "path.cpp",
+        "stacktrace.cpp",
+    ],
+    hdrs = glob([
+        "../../include/system/*.h",
+        "../../include/runtime/*.h",
+        "../../include/runtime/instance/*.h",
+    ]),
+    includes = [
+        "../../include",
+    ],
+    linkopts = select({
+        "@platforms//os:windows": ["-ldbghelp"],
+        "//conditions:default": [],
+    }),
+    deps = [
+        "//lib/common",
+    ],
+)
+
--- /dev/null
+++ b/lib/validator/BUILD.bazel
@@ -0,0 +1,31 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "validator",
+    srcs = glob([
+        "*.cpp",
+    ]),
+    hdrs = glob([
+        "../../include/validator/*.h",
+        "../../include/ast/*.h",
+        "../../include/ast/component/*.h",
+    ]),
+    includes = [
+        "../../include",
+    ],
+    deps = [
+        "//lib/common",
+        "//lib/loader",
+    ],
+)
+
--- /dev/null
+++ b/lib/executor/BUILD.bazel
@@ -0,0 +1,38 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "executor",
+    srcs = glob([
+        "*.cpp",
+        "engine/*.cpp",
+        "instantiate/*.cpp",
+        "instantiate/component/*.cpp",
+    ]),
+    hdrs = glob([
+        "../../include/executor/*.h",
+        "../../include/runtime/*.h",
+        "../../include/runtime/instance/*.h",
+        "../../include/ast/*.h",
+        "../../include/ast/component/*.h",
+    ]),
+    includes = [
+        "../../include",
+    ],
+    deps = [
+        "//lib/common",
+        "//lib/loader",
+        "//lib/system",
+    ],
+)
+
--- /dev/null
+++ b/lib/host/BUILD.bazel
@@ -0,0 +1,49 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "host_wasi",
+    srcs = glob([
+        "wasi/*.cpp",
+    ]) + select({
+        "@platforms//os:linux": glob(["wasi/linux/*.cpp"]),
+        "@platforms//os:macos": glob(["wasi/darwin/*.cpp"]),
+        "@platforms//os:windows": glob(["wasi/win/*.cpp"]),
+        "//conditions:default": [],
+    }),
+    hdrs = glob([
+        "../../include/host/*.h",
+        "../../include/host/wasi/*.h",
+    ]),
+    includes = [
+        "../../include",
+    ],
+    deps = [
+        "//lib/common",
+        "//lib/executor",
+    ],
+)
+
--- /dev/null
+++ b/lib/plugin/BUILD.bazel
@@ -0,0 +1,32 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "plugin",
+    srcs = glob([
+        "*.cpp",
+        "wasi_logging/*.cpp",
+    ]),
+    hdrs = glob([
+        "../../include/plugin/*.h",
+    ]),
+    includes = [
+        "../../include",
+    ],
+    deps = [
+        "//lib/common",
+        "//lib/executor",
+        "//lib/host:host_wasi",
+    ],
+)
+
--- /dev/null
+++ b/lib/vm/BUILD.bazel
@@ -0,0 +1,31 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "vm",
+    srcs = glob(["*.cpp"]),
+    hdrs = glob([
+        "../../include/vm/*.h",
+    ]),
+    includes = [
+        "../../include",
+    ],
+    deps = [
+        "//lib/common",
+        "//lib/executor",
+        "//lib/host:host_wasi",
+        "//lib/loader",
+        "//lib/plugin",
+        "//lib/validator",
+    ],
+)
+
--- /dev/null
+++ b/lib/po/BUILD.bazel
@@ -0,0 +1,26 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "po",
+    srcs = glob(["*.cpp"]),
+    hdrs = glob([
+        "../../include/po/*.h",
+    ]),
+    includes = [
+        "../../include",
+    ],
+    deps = [
+        "//lib/common",
+        "//lib/system",
+    ],
+)
+
--- /dev/null
+++ b/lib/api/BUILD.bazel
@@ -0,0 +1,34 @@
+# Copyright 2025 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+cc_library(
+    name = "wasmedge",
+    srcs = ["wasmedge.cpp"],
+    hdrs = glob([
+        "../../include/api/*.h",
+        "../../include/api/wasmedge/*.h",
+    ]) + [
+        "//include:copy_enum_headers",
+        "//include:generate_api_version_header",
+    ],
+    includes = [
+        "../../include",
+        "../../include/api",
+    ],
+    deps = [
+        "//lib/common",
+        "//lib/po",
+        "//lib/vm",
+    ],
+)
+
